src/project.rs

```diff
@@ -0,0 +1,659 @@
+        $crate::project!(&($c) $($f)*)
+    };
+    (&($c:expr) $($f:tt)*) => {
+        $crate::project!(
+            @inner
+            Borrow,
+            borrow,
+            project,
+            *const _,
+            inner,
+            &*inner,
+            addr_of,
+            $c,
+            $($f)*
+        )
+    };
+
+    (
+        @inner
+        $borrow_trait:ident,
+        $borrow_method:ident,
+        $project_fn:ident,
+        $ptr_ty:ty,
+        $inner_name:ident,
+        $convert_inner_raw_to_ref:expr,
+        $addr_of:ident,
+        $c:expr,
+        $($f:tt)*
+    ) => {{
+        // This function does nothing, but is unsafe to call, and so has the
+        // effect of requiring that the caller only invoke `project!` inside of
+        // an `unsafe` block.
+        $crate::project::promise_no_deref();
+        // We generate an `UnsafeToken` so that `$project_fn` can itself be
+        // safe, and thus we don't need to put the entire call to `$project_fn`
+        // in an `unsafe` block. This, in turn, is done so that the
+        // meta-variables `$c` and `$($f)*` are not expanded inside of an
+        // `unsafe` block, which would allow safe Rust code to smuggle in unsafe
+        // code via a call to `project!` without needing to write the `unsafe`
+        // keyword.
+        //
+        // TODO:
+        // - Slicing seems to be bounds-checked at runtime if need be, but is
+        //   this guaranteed by the reference/stdlib docs?
+        //
+        // SAFETY:
+        // - The arguments passed for `outer_to_inner` and
+        //   `field_to_wrapped_field` are just `as` casts, so they return
+        //   pointers to the same memory region as their arguments.
+        // - The argument passed for `inner_to_field` uses `addr_of!` (or
+        //   `addr_of_mut!`) to take the address of a sequence of field
+        //   accesses. By safety precondition of this macro, none of those field
+        //   accesses can go through a dereference. The only other types of
+        //   field accesses produce "places" which live entirely within the
+        //   memory region of the argument to `inner_to_field` as required by
+        //   the safety precondition of `project`/`project_mut`.
+        //
+        // Some older versions of Clippy have a bug in which they don't
+        // recognize the preceding safety comment.
+        #[allow(clippy::undocumented_unsafe_blocks)]
+        #[allow(unused_unsafe)]
+        let token = unsafe { $crate::project::unsafe_token::UnsafeToken::new() };
+        use ::core::borrow::$borrow_trait as _;
+        $crate::project::$project_fn(
+            token,
+            // We call `borrow` or `borrow_mut` so that this macro works
+            // regardless of whether `$c` is owned or borrowed.
+            //
+            // TODO(#196): Is there any way to make sure this calls
+            // `Borrow::borrow` or `BorrowMut::borrow_mut` even if the type has
```

Comment:
Not without spelling the type of `$c`.

`type alias impl Trait` could make this possible.

Reply:
Yeah, that's what I figured.

Also, could you leave comments for this module on https://github.com/google/zerocopy/pull/292? I hate GitHub...

Reply:
Oh wait, I just tried something. So long as the output type is non-ambiguous you can write `<_ as $borrow_trait>::$borrow_method($c)`. How does that work?

Reply:
dafuq...

I wonder if the compiler desugars them to the same thing (with the caveat that `$c.$borrow_method()` needs to also perform method resolution) and so they both get the same auto-ref treatment.

Reply:
In the `<_ as $borrow_trait>::$borrow_method($c)` you are _guaranteeing_ you're not running an inherent method and are running the trait method you intended, except you're leaving the concrete type that implements the trait as inferred.

Reply:
Marking this as resolved; we can [continue on the project PR](https://github.com/google/zerocopy/pull/292#discussion_r1309130530).